coderd/batchstats/batcher.go

```diff
@@ -0,0 +1,255 @@
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := database.Now()
+
+	b.buf.ID = append(b.buf.ID, uuid.New())
+	b.buf.CreatedAt = append(b.buf.CreatedAt, now)
+	b.buf.AgentID = append(b.buf.AgentID, agentID)
+	b.buf.UserID = append(b.buf.UserID, userID)
+	b.buf.TemplateID = append(b.buf.TemplateID, templateID)
+	b.buf.WorkspaceID = append(b.buf.WorkspaceID, workspaceID)
+
+	// Store the connections by proto separately as it's a jsonb field. We marshal on flush.
+	// b.buf.ConnectionsByProto = append(b.buf.ConnectionsByProto, st.ConnectionsByProto)
+	b.connectionsByProto = append(b.connectionsByProto, st.ConnectionsByProto)
+
+	b.buf.ConnectionCount = append(b.buf.ConnectionCount, st.ConnectionCount)
+	b.buf.RxPackets = append(b.buf.RxPackets, st.RxPackets)
+	b.buf.RxBytes = append(b.buf.RxBytes, st.RxBytes)
+	b.buf.TxPackets = append(b.buf.TxPackets, st.TxPackets)
+	b.buf.TxBytes = append(b.buf.TxBytes, st.TxBytes)
+	b.buf.SessionCountVSCode = append(b.buf.SessionCountVSCode, st.SessionCountVSCode)
+	b.buf.SessionCountJetBrains = append(b.buf.SessionCountJetBrains, st.SessionCountJetBrains)
+	b.buf.SessionCountReconnectingPTY = append(b.buf.SessionCountReconnectingPTY, st.SessionCountReconnectingPTY)
+	b.buf.SessionCountSSH = append(b.buf.SessionCountSSH, st.SessionCountSSH)
+	b.buf.ConnectionMedianLatencyMS = append(b.buf.ConnectionMedianLatencyMS, st.ConnectionMedianLatencyMS)
+
+	// If the buffer is full, signal the flusher to flush immediately.
+	if len(b.buf.ID) == cap(b.buf.ID) {
+		b.flushLever <- struct{}{}
+	}
+	return nil
+}
+
+// Run runs the batcher.
+func (b *Batcher) Run(ctx context.Context) {
+	// nolint:gocritic
+	authCtx := dbauthz.AsSystemRestricted(ctx)
+	for {
+		select {
+		case <-b.ticker:
+			b.flush(authCtx, false, "scheduled")
+		case <-b.flushLever:
+			// If the flush lever is depressed, flush the buffer immediately.
+			b.flush(authCtx, true, "full buffer")
+		case <-ctx.Done():
+			b.log.Warn(ctx, "context done, flushing before exit")
+			b.flush(authCtx, true, "exit")
+			return
+		}
+	}
+}
+
+// flush flushes the batcher's buffer.
+func (b *Batcher) flush(ctx context.Context, forced bool, reason string) {
+	b.mu.Lock()
+	start := time.Now()
+	count := len(b.buf.ID)
+	defer func() {
+		b.mu.Unlock()
+		// Notify that a flush has completed.
+		if b.flushed != nil {
+			b.flushed <- forced
+		}
+		if count > 0 {
+			elapsed := time.Since(start)
+			b.log.Debug(ctx, "flush complete",
+				slog.F("count", count),
+				slog.F("elapsed", elapsed),
+				slog.F("forced", forced),
+				slog.F("reason", reason),
+			)
+		}
+	}()
+
+	if len(b.buf.ID) == 0 {
+		return
+	}
+
+	// marshal connections by proto
+	payload, err := json.Marshal(b.connectionsByProto)
+	if err != nil {
+		b.log.Error(ctx, "unable to marshal agent connections by proto, dropping data", slog.Error(err))
+		b.buf.ConnectionsByProto = json.RawMessage(`[]`)
+	} else {
+		b.buf.ConnectionsByProto = payload
+	}
+
+	err = b.store.InsertWorkspaceAgentStats(ctx, b.buf)
```

Comment:
What happens if one "stat" in the batch is faulty? Is the whole batch rejected?

Bonus points: maybe we need a test for this.

Reply:
Yes, the whole batch will be rejected. That's a good idea for a test.

Reply:
It would be nice to exclude the faulty item, and keep the rest of the batch, but I guess that it would be tricky to implement, or just fall back to iterative approach.